### PR TITLE
Convert switch arms with their own binder

### PR DESCRIPTION
**Pull request by alrz**
Fixes https://github.com/dotnet/roslyn/issues/54193

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 1]
```diff
@@ -340,10 +340,11 @@ private BoundExpression ConvertSwitchExpression(BoundUnconvertedSwitchExpression
             {
                 var oldCase = source.SwitchArms[i];
                 var oldValue = oldCase.Value;
+                var binder = GetRequiredBinder(oldCase.Syntax);
                 var newValue =
                     targetTyped
-                    ? CreateConversion(oldValue.Syntax, oldValue, underlyingConversions[i], isCast: false, conversionGroupOpt: null, destination, diagnostics)
-                    : GenerateConversionForAssignment(destination, oldValue, diagnostics);
+                    ? binder.CreateConversion(oldValue.Syntax, oldValue, underlyingConversions[i], isCast: false, conversionGroupOpt: null, destination, diagnostics)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:346]: I couldn't follow how using a different binder for conversion could result in the original `The name 'n' does not exist in the current context` error (which doesn't seem related to conversion). How does this work?

Comment (alrz) [reply src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:346]: The switch expression is the only expression that introduces locals in its own scope, so we need the actual binder of the switch arms here to get access to those scoped locals while binding `new()` initializer which happens during the conversion.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 2)

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 2]
```diff
@@ -340,10 +340,11 @@ private BoundExpression ConvertSwitchExpression(BoundUnconvertedSwitchExpression
             {
                 var oldCase = source.SwitchArms[i];
                 var oldValue = oldCase.Value;
+                var binder = GetRequiredBinder(oldCase.Syntax);
```
Comment (alrz) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:343]: ❓ I'm not sure if this is the right way to acquire the binder. Should we capture it in the bound node instead?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:343]: >GetRequiredBinder(oldCase.Syntax) [](http://example.com/codeflow?start=29&length=33)

It feels like this code will be correct only when ```oldCase.Syntax``` is a ```SwitchExpressionArmSyntax```. Please add an assert.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ImplicitObjectCreationTests.cs [hunk 3]
```diff
@@ -4659,5 +4659,27 @@ public class C : System.Attribute
...
+Write(c1.n);
+Write(c2.n);
+
+class C 
+{
+    public int n;
+}
+";
+            var compilation = CreateCompilation(source);
+            compilation.VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ImplicitObjectCreationTests.cs:4681]: >VerifyDiagnostics [](http://example.com/codeflow?start=24&length=17)

Consider calling this on the result of ```CompileAndVerify``` instead.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 4)

**Comment by AlekseyTs** [comment]
@alrz Thanks for the contribution.
